state/upgrades.go

```diff
@@ -2543,3 +2543,52 @@ func ReplacePortsDocSubnetIDCIDR(pool *StatePool) (err error) {
 		return nil
 	}))
 }
+
+// EnsureRelationApplicationSettings creates an application settings
+// doc for each endpoint in each relation if one doesn't already
+// exist.
+func EnsureRelationApplicationSettings(pool *StatePool) error {
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		settingsCol, closer := st.db().GetCollection(settingsC)
+		defer closer()
+
+		allSettings := set.NewStrings()
+		settingsIter := settingsCol.Find(nil).Iter()
+		defer settingsIter.Close()
+
+		var doc struct {
+			ID string `bson:"_id"`
+		}
+		for settingsIter.Next(&doc) {
+			allSettings.Add(doc.ID)
+		}
+		if err := settingsIter.Close(); err != nil {
+			return errors.Trace(err)
+		}
+
+		relations, err := st.AllRelations()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var ops []txn.Op
+		for _, rel := range relations {
+			if rel.Life() != Alive {
```

Comment:
Are we sure that not having settings for Dying relations is ok everywhere? 
Isn't it safer to create always and let the relation remove code get rid of them when the relation is eventually deleted?

Reply:
I have this weird feeling that there was a reason that I added it, but I can't remember it now. :( I'll remove it, hopefully whatever the reason was becomes apparent.